### PR TITLE
LibJS/JIT: Fast path for decrement & better opcodes

### PR DESCRIPTION
**Pull request by feliwir**
- [x] Use `inc` instruction for our existing increment fastpath
- [x] Add a decrement fast path that uses the `dec` instruction
- [x] Make integer limits check more obvious by using `INT32_MAX`   / `INT32_MIN` macros

### CODE REVIEW

File: Userland/Libraries/LibJS/JIT/Compiler.cpp [hunk 1]
```diff
@@ -404,9 +403,45 @@ static Value cxx_decrement(VM& vm, Value value)
...
+        // GPR0 = ARG1 & 0xffffffff;
+        m_assembler.mov(
+            Assembler::Operand::Register(GPR0),
+            Assembler::Operand::Register(ARG1));
+        m_assembler.mov(
+            Assembler::Operand::Register(GPR1),
+            Assembler::Operand::Imm(0xffffffff));
+        m_assembler.bitwise_and(
+            Assembler::Operand::Register(GPR0),
+            Assembler::Operand::Register(GPR1));
```
Comment (skyrising) [inline Userland/Libraries/LibJS/JIT/Compiler.cpp:420]:
```suggestion
        m_assembler.mov32(
            Assembler::Operand::Register(GPR0),
            Assembler::Operand::Register(ARG1));
```

**Comment by awesomekling** [comment]
As conventional wisdom says `inc` and `dec` are slower due to partial flag update stalls, please provide data supporting `inc reg` and `dec reg` being faster than `add reg, 1` and `sub reg, 1` in our context.

File: Userland/Libraries/LibJS/JIT/Compiler.cpp (hunk 1 above)
Comment (skyrising) [reply Userland/Libraries/LibJS/JIT/Compiler.cpp:420]: This might need an `Assembler::Extension::SignExtend`. Please add a test for this edge-case to verify it works.

File: Userland/Libraries/LibJIT/X86_64/Assembler.h [hunk 2]
```diff
@@ -742,6 +742,17 @@ struct X86_64Assembler {
         }
     }
 
+    void increment(Operand op)
+    {
+        if (op.is_register_or_memory()) {
+            emit_rex_for_OI(op, REX_W::Yes);
```
Comment (awesomekling) [inline Userland/Libraries/LibJIT/X86_64/Assembler.h:748]: `OI` is "Opcode + Immediate" (according to @Hendiadyoin1), and since there's no immediate here, we should probably use `emit_rex_for_slash()` instead :)

Same for the decrement change.

**Comment by skyrising** [comment]
> We had that discussion yesterday in #development (Discord) and came to the conclusion that they're equally fast.

I wouldn't really say we "came to a conclusion". We found a source claiming something, but we should still verify whether that's true.

File: Userland/Libraries/LibJS/JIT/Compiler.cpp [hunk 3]
```diff
@@ -367,11 +367,11 @@ void Compiler::compile_increment(Bytecode::Op::Increment const&)
...
             Assembler::Operand::Register(GPR1));
 
-        // if (GPR0 == 0x7fffffff) goto slow_case;
+        // if (GPR0 == INT32_MAX) goto slow_case;
         m_assembler.jump_if(
             Assembler::Operand::Register(GPR0),
             Assembler::Condition::EqualTo,
-            Assembler::Operand::Imm(0x7fffffff),
+            Assembler::Operand::Imm(INT32_MAX),
             slow_case);
```
Comment (awesomekling) [inline Userland/Libraries/LibJS/JIT/Compiler.cpp:375]: Please do this in a separate commit, as it has nothing to do with adding a decrement fast path.

File: Userland/Libraries/LibJIT/X86_64/Assembler.h [hunk 4]
```diff
@@ -742,6 +742,36 @@ struct X86_64Assembler {
         }
     }
 
+    void inc32(Operand op, Optional<Label&> overflow_label)
+    {
+        if (op.is_register_or_memory()) {
+            emit_rex_for_slash(op, REX_W::Yes);
```
Comment (skyrising) [inline Userland/Libraries/LibJIT/X86_64/Assembler.h:748]:
```suggestion
            emit_rex_for_slash(op, REX_W::No);
```

**Comment by awesomekling** [comment]
I'm unconvinced about the `inc` vs `add` thing, but sure let's take it as a code size optimization. We can sweat the instruction selection in the future when we run out of low-hanging fruit :)
